Change FFDShow DPL2 decoder to FreeSurround

LAGonauta opened the pull request:
Added the FreeSurround decoder to the mixer, and removed the legacy FFDShow decoder.
The decoding is done through the new SurroundDecoder class using a FixedSizeQueue as a ring buffer as the decoder needs a fixed amount of frames to be able to decode the surround information.

Tested on XAudio2 (with PR #6171), OpenAL, and Cubeb on Windows.
Also tested on OpenAL, Cubeb, and PulseAudio on Ubuntu, but Cubeb seems to be downmixing the surround information to stereo. Works on Windows, so maybe there is a bug in Cubeb library associated with Linux.

The upmix works really well, but I am not sure if we should set SURROUND_BLOCK_SIZE to 1024, 512, or even 256. All work, but the algorithm has more quality with 1024, less steering glitches and less crosstalk. For now I set it to 512 for an OK mix of quality and latency. The best quality is with SURROUND_FRAMES_PER_CALL set to 4096, but the latency seems to be equivalent to time-stretching with a buffer size of 80 ms.

Fixes [bug #9047](https://bugs.dolphin-emu.org/issues/9047).

~The new decoder seems to have some kind of incompatibility with some distros (Ubuntu 16.04, for example) as it causes glitches on decode. I will investigate the glitches later.~
**It seems that a fully updated Ubuntu 16.04 distro does not show this behaviour anymore.**

Any suggestion is greatly appreciated :)

lioncash commented:
All of this external library code should likely be placed into Externals, rather than AudioCommon, especially the FreeSurround stuff.

LAGonauta commented:
Just moved everything to Externals.

Orphis commented:
Where is the CMakeLists.txt for non-Visual Studio platforms?

Orphis commented on CMakeLists.txt:
```diff
@@ -657,8 +657,10 @@ if(OPENAL_FOUND)
   endif()
+  message(STATUS "Using static FreeSurround from Externals")
+  add_subdirectory(Externals/FreeSurround)
+  include_directories(Externals)
```
Remove this include_directories(), this is a call that adds the include path to ALL the targets.

Orphis commented on Externals/FreeSurround/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+	FreeSurroundDecoder.cpp
+	)
+
+add_library(FreeSurround STATIC ${SRCS})
```
Put: target_include_directories(FreeSurround PUBLIC .) here to do the same as before, but better.

BUT! Using just "Externals" means you could include any other headers from that folder by linking this target and propagating the include directories. This breaks isolation.

It also means that all the other includes should be changed from "FreeSurround/foo.h" to just "foo.h". Alternatively, you could put the public API headers in a folder "Externals/FreeSurround/include/FreeSurround" and just add the folder "include" and keep the same include directives you currently have.

LAGonauta replied:
I did not change this yet as I was thinking how I should also configure Visual Studio to make it work on both. Will do later.

LAGonauta replied:
Should be all good now :)

JMC47 commented:
@dolphin-emu-bot rebuild

LAGonauta commented:
I noticed that I had forgotten to update the References on the Debug build for Visual Studio. Now everything should compile fine there.

I am doing experiments on padding the values to not only be able to use SoundTouch but also to allow its use on other backends. I noticed that PulseAudio, for example, rarely returns the desired 512 samples, it oscillates over it (505-520).
Sadly still no go, you can see my "progress" here: https://github.com/LAGonauta/dolphin/tree/fs-buffer-tests

Even though some days ago SoundTouch had horrible latency when forcing the output of 512 samples (for me), I found out the latency is bearable when changing the default periods and period size on OpenAL Soft's config file to small numbers like 3 and 128, respectively. Maybe on the OpenAL backend we can just do that? Personally I prefer to use OpenAL's frequency-shifter.

LAGonauta commented:
Just rebased this PR with the changes from merry's audio-stretching PR.
Now just need to find a way to make the buildbot not get stuck while trying to compile...

ligfx commented:
@dolphin-emu-bot rebuild

BhaaLseN commented:
Nothing you can do on your end, buildbot has a size limit on patches it can receive, and it seems our monkey-patch for that got lost in the upgrade. delroth knows, but he's on vacation and probably won't work on it until he's back.

BhaaLseN commented on CMakeLists.txt:
```diff
@@ -650,6 +650,10 @@ else()
 endif()
 
 find_package(OpenAL)
+if(OPENAL_FOUND)
```
The other definitions usually print "using from Externals" if _no_ system library was found. Did you mean `NOT OPENAL_FOUND`, or do you mean to _always_ use the one from Externals?

LAGonauta replied:
Currently FreeSurround only works on the OpenAL backend, this is why I made it look for OpenAL. There is no other reason.
It always use from externals, as I believe there will never be a system-wide FreeSurround library available.

ligfx commented:
@dolphin-emu-bot rebuild

Orphis commented on Externals/FreeSurround/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+
+add_library(FreeSurround STATIC ${SRCS})
+target_include_directories(FreeSurround PUBLIC include)
+add_definitions(-w)
```
target_compile_options(FreeSurround PRIVATE -w)

Orphis commented on Externals/FreeSurround/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+	source/KissFFT.cpp
+	source/KissFFTR.cpp
+	source/FreeSurroundDecoder.cpp
+	)
```
Indentation is 2 spaces, closing paren should be aligned with set (I know we are inconsistent for that one)

spycrab commented:
@dolphin-emu-bot rebuild

ligfx commented:
@dolphin-emu-bot rebuild

LAGonauta commented:
Just rebased because of Cubeb. I will also try to add support for FreeSurround on Cubeb, but right now it is not respecting how many frames I ask for.

ligfx commented:
@dolphin-emu-bot rebuild

ligfx commented:
@dolphin-emu-bot rebuild

ligfx commented:
@LAGonauta I said a few things on IRC before I realized you were gone :)

> ligfx 7:11pm: LAGonauta does it alway ask for exactly 480 frames? is there maybe one callback at the beginning that asks for the full amount you requested (512)?
> ligfx 7:13pm: I'm thinking it might be correctly honoring the request for a larger buffer, but then the callbacks get scheduled at the same time as the audio engine (every 480 frames), which leaves 32 frames unread / only 480 writable frames

shuffle2 commented on Externals/FreeSurround/FreeSurround.vcxproj:
```diff
@@ -0,0 +1,66 @@
+  <PropertyGroup Label="UserMacros" />
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <Lib>
+      <AdditionalLibraryDirectories>%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
```
Please have a look at `Source\VSProps\Base.props` for reference as to how includes are added.
I'm not sure why people started putting the `AdditionalLibraryDirectories` in project's own .vcxproj, recently...

(Note: this behavior was to match the cmake behavior at the time I overhauled the VS projects when moving to vs2015...it would be interesting to know if cmake behavior has changed)

LAGonauta replied:
Will take a look. This file was created automatically by VS2015 when I first imported the library, maybe it added some "junk" while doing so.

LAGonauta replied:
Done, I hope.

shuffle2 commented on Source/Core/AudioCommon/AudioCommon.vcxproj:
```diff
@@ -78,6 +88,9 @@
     <ProjectReference Include="$(CoreDir)Common\Common.vcxproj">
       <Project>{2e6c348c-c75c-4d94-8d1e-9c1fcbf3efe4}</Project>
     </ProjectReference>
+    <ProjectReference Include="..\..\..\Externals\FreeSurround\FreeSurround.vcxproj">
```
you can use `$(CoreDir)` here.

LAGonauta replied:
Done.

shuffle2 commented on Source/Core/AudioCommon/AudioCommon.vcxproj:
```diff
@@ -35,6 +35,16 @@
   <PropertyGroup Label="UserMacros" />
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <ClCompile>
+      <AdditionalIncludeDirectories>$(ExternalsDir)FreeSurround\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
same comment re: includes

LAGonauta replied:
Done, I hope.

LAGonauta commented:
@ligfx No, the first callback asks for 960 frames while all other callback asks for 480 (may depend on the CPU usage, I guess). No callback ever asks for 512 frames.

shuffle2 commented:
lgtm wrt previous comments

I'm curious if this is removing openal? (since you removed `find_package(OpenAL)`).

We seem to be acquiring a large variety of 3rd party audio dependencies...

LAGonauta commented:
Oops, this is not removing OpenAL, even though now it is not very useful on non-Windows platforms with Cubeb and with time-stretching integrated in the mixer (Windows at least has low-latency audio and high quality resamplers on sound cards with hardware OpenAL).

I did not mean to delete this line, fixed now.

ligfx commented:
@LAGonauta I think I found the WASAPI issue: https://github.com/kinetiknz/cubeb/issues/324

ligfx commented:
> Just tested on Ubuntu and the audio seems to be glitching, the last 2-3 samples of each buffer are getting set to zero for no apparent reason... Just tested it again on Windows and everything is working as it should. Maybe the compiler on Linux is generating bad code? Can someone test on a Mac, please?

No glitching for me on macOS.

> Note that it seems that Cubeb on Windows can output 5.1 only when the sample rate is set to 48kHz on the device control panel.
>
> I also added the surround support to the XAudio2 backend, but only works when 5.1 is set on the audio device control panel.

What's up with the restrictions on the Windows backends? Can those be fixed?

> I am also not using FreeSurround's source code directly. For example, I added a function to change the BlockSize and another to change the SampleRate, both on runtime.

Please separate out "adding FreeSurround source code" and "Dolphin-specific code changes" into separate commits. What's the dynamic block size used for? The sample rate should not need to be changed dynamically, since a `Mixer` instance sticks to a single sample rate (looks like the freesurround instance is a static reference, it should probably a member of `Mixer`).

LAGonauta commented:
Just split my first commit into adding FS to Externals and adding it to the mixer.
Also removed the dynamic block size and sample rate, really not needed.

This is probably a bug in Cubeb, I need to investigate more.
About XAudio2, it seems to be a old bug that was never (and will never) be fixed. I will investigate if we can workaround it without hex-editing files and reg-editing (seems to be the common "solution")

About making FreeSurround a member of Mixer, probably. I am not sure how to do that and will push a commit later on after studying more.

leoetlino commented:
The includes should be of the form `#include <...>` for FreeSurround stuff, not `#include "..."`.

The build failures are due to the fact AudioCommon is used from Core, but the dependency on FreeSurround is private to AudioCommon and not exposed to Core (as far as I can tell). You can fix this by adding a dependency on FreeSurround in the Core CMakeLists.txt explicitly (like how it's done for cubeb). There's probably a better way to do this, but I don't know much about cmake stuff.

MerryMage commented:
> You can fix this by adding a dependency on FreeSurround in the Core CMakeLists.txt explicitly (like how it's done for cubeb). There's probably a better way to do this, but I don't know much about cmake stuff.

     target_link_libraries(audiocommon PUBLIC FreeSurround PRIVATE cubeb SoundTouch)

LAGonauta commented:
Thanks, I added `target_link_libraries(audiocommon PUBLIC FreeSurround PRIVATE cubeb SoundTouch)` and now it is compiling on very platform :)
Maybe @Orphis could chime in if this is the most desirable solution? Thanks.

FreeSurround is now a member of CMixer (it is, right?), but I still couldn't find out what is causing the glitches when I compile on Ubuntu... I guess I will need to learn how to use GDB with Dolphin.
One thing is for sure: they are generated somewhere within FreeSurround, before reaching the mixer.

Orphis commented:
If FreeSurround is part of the public interface of audiocommon, then that is indeed a solution.
Another one would be not to expose it if it's an implementation detail.

LAGonauta commented:
I tried to debug it with GDB but couldn't find out why there are glitches when decoding on Ubuntu, while it decodes perfectly on Windows and MacOS... If someone with more experience could take a look I would be thankful.
There is no need to have a full 5.1 surround system as PulseAudio/Cubeb should remix it to stereo after decoding

LAGonauta commented:
Just tested on Ubuntu 17.04. No glitches, perfect on any backend with surround support.
Maybe it is a bug in GCC/Clang (doubt it) or in a library (maybe libc++ or something) that was fixed on later revisions.

How many people are using Ubuntu 16.04? Does Dolphin collects that information?

LAGonauta commented:
Instead of completely removing the old decoder I choose to add an hidden option to select which one to use. People on incompatible OS's won't feel left out, and everybody else will be able to use the far better FreeSurround decoder.

Any other suggestion will be greatly appreciated :)

fantesykikachu commented:
well there are a couple options:
- what you you have right not
- we only build the new decoder on windows and macos and then on linux build the old one
- adding a cmake flag to use the old decoder (this causes problems for the community mangers)
- someone takes a copy of ubuntu 16.04 and manually upgrade files until it is fixed and then check it is not ubuntu specific(as they modify a lot of things) then either:
   - select which decoder to build based on that 
   - have a cmake error force them to upgrade what ever files necessary ( this puts a lot of work on the end user)

personally i don't like hidden options or cmake flags as they need to be informed about them. i think the last option is the best especially if we expect most users to use repos or the ppa since we already have a notice on the installing dolphin page telling ubuntu 14.04 users to upgrade libstdc++6.

ligfx commented:
Wow, I finally got a chance to compare this directly against the old decoder (shout-out to to #6129 for making it easy), and this sounds so so so much better! The old decoder really distorts the audio output.

I think the next steps on this would be:
- Extract the meat of the MixSurround logic into a separate class, like the existing AudioCommon AudioStretcher.
- Add some sort of buffer/FIFO to smooth over mismatched chunk sizes. (E.g. if PulseAudio asks for 505 frames, we should get 512 from the mixer and keep the extra 7 frames for the next callback. Does that make sense?)
- Remove the old decoder entirely. If it glitches on some systems, then we should fix that, but I don't see a point in keeping the code around when it sounds so terrible.
- Split out the changes that add surround sound support in the XAudio backends. It's cool stuff, and can probably get merged faster on its own.

LAGonauta commented:
@ligfx 
Just removed the XAudio2 changes from this commit and created another PR ;)

About the older decoder:
I agree, but I wasn't able to fix it by myself... I really need to learn GDB, I am too used with Visual Studio. Also, as it works on Ubuntu 17.04 it is difficult for me to find the will to do so... haha

About the FIFO:
I added a small FIFO like that on Mixer.cpp using std:queue, starting on line 164. It works fine but it is not perfect. If you have any suggestion on how to make it better I will definitely look into it.

About extracting the meat:
Will look into that, I will probably use the AudioStretcher class as base.

Orphis commented:
@LAGonauta You can try Visual Studio Code, it gives a good experience on non Windows platforms

LAGonauta commented:
@ligfx Updated the PR with your suggestions, but I still think there is a better design for the SurroundDecoder class than the one I made... Any suggestions?

ligfx commented:
@LAGonauta I'd suggest not passing the Mixer instance into the surround decoder: it should be able to be standalone (principle of decreased coupling).

You might be interested in an API similar to [SoundTouch](https://github.com/dolphin-emu/dolphin/blob/master/Externals/soundtouch/SoundTouch.h): a `PutSamples(ptr, size)` method to add unprocessed data, a `size_t NumSamples()` method to query the available processed data, and a `ReceiveSamples(ptr, size)` method to receive processed data.

Add to that some way to query the minimum number of frames needed to increase NumSamples (maybe like, `size_t QueryNeededInputForOutput(size_t)`? Ugly name, but hopefully you get the idea), and then you could do something like:

```c++
std::size_t SurroundDecoder::QueryNeededInputForOutput(std::size_t output_samples)
{
  return output_samples + (SURROUND_FRAMES_PER_CALL - output_samples % SURROUND_FRAMES_PER_CALL);
}

void Mixer::MixSurround(float* samples, unsigned int num_samples)
{
  size_t needed_samples = m_decoder.QueryNeededInputForOutput(num_samples);
  // Mix() may also use m_scratch_buffer internally, but is safe because it alternates reads 
  // and writes.
  unsigned int available_samples = Mix(m_scratch_buffer.data(), needed_samples);
  if(available_samples != needed_samples) this_is_bad();

  m_decoder.PutSamples(m_scratch_buffer.data(), needed_samples)
  m_decoder.ReceiveSamples(samples, num_samples);
  return num_samples;
}
```

Nits:
- `Mixer::m_decoder` is a bit ambiguous, what about something like `m_surround_decoder`?
- Earlier there was some discussion about the need to make FreeSurround a public CMake dependency of AudioCommon: if you don't want that, an easy thing to do now would be to change `SurroundDecoder::m_fsdecoder` to a unique_ptr (the PImpl idiom, sort of).

LAGonauta commented:
@ligfx 
I liked your idea for the API, just pushed the changes.
But I couldn't remove the need of a PUBLIC FreeSurround yet. I will study more about the PImpl idiom and then try later.

LAGonauta commented:
I managed to make FreeSurround private by using a shared_ptr on SurroundDecoder.h, but something doesn't feel right...
When I try to use a unique_ptr the GCC compiler doesn't like it, maybe because of something about the destructor not existing. Anyway, the actual error is about the size of an incomplete type not being available.

MerryMage commented:
> When I try to use a unique_ptr the GCC compiler doesn't like it, maybe because of something about the destructor not existing. Anyway, the actual error is about the size of an incomplete type not being available.

If you do not provide a destructor for `SurroundDecoder`, the compiler generates one for your automatically. This automatically generated destructor requires the complete declaration of `DPL2FSDecoder` and not just a forward reference.

You can fix this by declaring a destructor in SurroundDecoder.h:

        ~SurroundDecoder();

And implementing it in SurroundDecoder.cpp:

    SurroundDecoder::~SurroundDecoder() = default;

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.h:
```diff
@@ -0,0 +1,36 @@
+class SurroundDecoder
+{
+public:
+  explicit SurroundDecoder(uint32_t sample_rate, uint32_t frame_block_size);
```
We generally use our `u32` type alias instead of `uint32_t` (ditto for any other `<s|u>[size]_t` types)

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.cpp:
```diff
@@ -0,0 +1,92 @@
+void SurroundDecoder::PutFrames(short* in, size_t num_frames_in)
+{
+  // Maybe check if it is really power-of-2?
+  int remaining_frames = static_cast<int>(num_frames_in);
```
why cast to an `int` as opposed to a 64-bit type like `s64` (since size_t is 64-bits on most platforms we target)?

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.cpp:
```diff
@@ -0,0 +1,92 @@
+
+namespace AudioCommon
+{
+SurroundDecoder::SurroundDecoder(uint32_t sample_rate, uint32_t frame_block_size)
```
An initializer list can be utilized here.

```cpp
SurroundDecoder::SurroundDecoder(uint32_t sample_rate, uint32_t frame_block_size)
    : m_sample_rate{sample_rate}, m_frame_block_size{frame_block_size} 
{
  m_fsdecoder = std::make_shared<DPL2FSDecoder>();
  m_fsdecoder->Init(cs_5point1, m_frame_block_size, m_sample_rate);
}
```

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.cpp:
```diff
@@ -0,0 +1,92 @@
+        output_frames - static_cast<uint32_t>(m_decoded_fifo.size()) / SURROUND_CHANNELS;
+    return frames_needed + m_frame_block_size - frames_needed % m_frame_block_size;
+  }
+  else
```
`else` can be omitted, since the `if` contains a return.

i.e.

```cpp
if (...)
{
  return ...;
}
else
{
  return ...;
}
```

can be

```cpp
if (...)
{
  return ...;
}

return ...;
```

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.h:
```diff
@@ -0,0 +1,36 @@
+{
+public:
+  explicit SurroundDecoder(uint32_t sample_rate, uint32_t frame_block_size);
+  uint32_t QueryFramesNeededForSurroundOutput(uint32_t output_frames);
```
This can be a const member function.

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.cpp:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    // Decode
+    float* dpl2_fs = m_fsdecoder->decode(m_float_conversion_buffer.data());
```
As far as I can tell, this can be a `const float*`

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.cpp:
```diff
@@ -0,0 +1,92 @@
+    // FL | FR | FC | LFE | BL | BR
+    for (size_t i = 0; i < m_frame_block_size; ++i)
+    {
+      m_decoded_fifo.push(dpl2_fs[i * SURROUND_CHANNELS + 0 /*LEFTFRONT*/]);
```
Could these comments not be put outside of the code itself, like:

```cpp
m_decoded_fifo.push(dpl2_fs[i * SURROUND_CHANNELS + 0]);  // LEFTFRONT
```
?

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.cpp:
```diff
@@ -0,0 +1,92 @@
+}
+
+// Receive and decode samples
+void SurroundDecoder::PutFrames(short* in, size_t num_frames_in)
```
As far as I can tell, this can be a `const short*`

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.cpp:
```diff
@@ -0,0 +1,92 @@
+#include <limits>
+
+#include "AudioCommon/SurroundDecoder.h"
+#include "FreeSurround/FreeSurroundDecoder.h"
```
If this is referencing an external, this should be using `<>` instead of `""`

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.h:
```diff
@@ -0,0 +1,36 @@
+  std::shared_ptr<DPL2FSDecoder> m_fsdecoder;
+  const uint32_t STEREO_CHANNELS = 2;
+  const uint32_t SURROUND_CHANNELS = 6;
+  std::array<float, 32768> m_float_conversion_buffer;
```
`<array>` needs to be included.

lioncash commented on Source/Core/AudioCommon/SurroundDecoder.h:
```diff
@@ -0,0 +1,36 @@
+  uint32_t m_frame_block_size;
+
+  std::shared_ptr<DPL2FSDecoder> m_fsdecoder;
+  const uint32_t STEREO_CHANNELS = 2;
```
these constants can likely just be `constexpr u32`s in the cpp file.

LAGonauta commented:
@lioncash 
I made all changes you requested.

@MerryMage 
Oh, that was easy. I was looking for the problem on DPL2FSDecoder itself, not on the SurroundDecoder class! Thanks for the heads up. I need to study more how PImpl works.

JMC47 commented:
Is this Pull Request still active?

LAGonauta commented:
I am not sure when I will be able to track why the audio is glitching on Ubuntu 16.04 as it seems to be some incompatibility between FreeSurround/KissFFT with a library from the system and currently my time is short.

I would make changes on this PR only if a reviewer requested, and I vote merging even with the glitches as Ubuntu 18.04 is around the corner and the current decoder might even be considered broken everywhere anyway.

LAGonauta commented:
Anything I can do to make this PR move forward for merge?

LAGonauta commented:
After extensive testing I found that it is still needed to clamp some rare samples depending on the game, re-added that function for better user experience.

Also rebased the branch to the current master.

LAGonauta commented:
Recently had the time try to find out what caused the glitches on Ubuntu 16.04 and they seem to have been fixed when using a fully updated distro. Yay! I guess?

shuffle2 commented:
This PR seems fine to me, tbh my guess would be that people aren't merging it just because not many people have a way to test it...

weihuoya commented:
it's work fine on android, thx.

DanLaRoche commented:
It's a shame that the reviewers are ignoring this PR, some of us would love to get surround sound working correctly without having to rely on inconsistent pro logic support on modern HTIB systems and sound bars. In the meantime, LAGonauta, could you do a pull from master to keep your fork up to date? I'll be testing this soon on my HTPC that is connected to a surround sound system.

LAGonauta commented:
@DanLaRoche Just rebased and pushed.
Remember to use Cubeb or OpenAL on Windows and PulseAudio on Linux. Other backends do not support surround sound yet.

stenzek commented:
The code seems fine to me, aside from a few small nitpicks (e.g. replace the clamping by if/else with MathUtil::Clamp). But I'm not very familiar with the audio side of Dolphin, so I'm probably not the best person to review it.

But it's a shame for this to be sitting around, so unless anyone has any better ideas, I could merge it as-is and deal with any regressions later.

Zexaron commented:
@LAGonauta Needs another rebase. I have a 5.1 setup my self and I'm sad at the gaming/PC HW audio industry, audio is extremely underappreciated, so my hat off for this PR.

LAGonauta commented:
@ghost 
Just rebased.

retrue commented:
Hi
I have noticed problems with this change.
I am playing Resident Evil 4, with Surround activated. Before this pull request, sound was perfect. Now, it is choppy. If I disable Dolby Pro Logic II Decoder in Sound Settings, the sound is clean again.
My computer; i7 4770, 16 GB RAM, Nvidia 1060 6GB, SSD. Integrated sound chip Realtek High Definition Audio. OS, Windows 10 Pro 64bit.
My sound settings, in the image.
![soundconfigdolby](https://user-images.githubusercontent.com/18194241/52780684-108f0680-304b-11e9-80a6-cf3cca59115c.png)

retrue commented:
Hi, I have made more tests. With the above settings, using HLE and LLE Interpreter I get bad sound too.
But if I disable Audio Stretching, sound is again good with Dolby Pro Logic enabled.

LAGonauta commented:
Investigated a little bit and it seems this is happening and it seems that the non-uniform number of samples output from SoundTouch might be doing that... I need to investigate more.
However, what do you mean with not as good as before the PR without audio stretching? Is it stuttering for you? Or just different?
The older decoder did not really do any decoding, so with FreeSurround the sound placement will be really different.

retrue commented:
Hi
What I meant is that  it seems FreeSurround and audio stretching are not working well together and produces bad sound.
If I disable audio stretching the audio improves a lot but there is a little bit of stuttering. My opinion (I can be wrong) is that you are simply losing the fluidity in the sound you would get with audio stretching enabled.
